File: leetcode/practice/695-max-area-of-island/dfs.cpp
class Solution {
public:
    int maxArea ;
    int startRow , startCol ;
    int maxAreaOfIsland(vector<vector<int>>& grid) ;
    int dfs(vector<vector<int>>& grid , vector<vector<bool>>& visited) ;
    bool validMove(int currRow , int currCol , int row , int col) ;
};
bool Solution :: validMove(int currRow , int currCol , int row , int col)
{
    return currRow>=0 and currCol>= 0 and currRow < row and currCol < col ;
}
int Solution :: dfs(vector<vector<int>>& grid ,vector<vector<bool>>&visited)
{
    stack<pair<int , int>>myStack ;
    myStack.push({startRow , startCol}) ;
    visited[startRow][startCol] = true ;
    int area = 0 ;
    while(!myStack.empty())
    {
        auto info = myStack.top() ;
        myStack.pop() ;
        int currRow = info.first , currCol = info.second ;
        area++ ;
        if(validMove(currRow - 1 , currCol , grid.size() , grid[0].size()))
        {
            if(grid[currRow - 1][currCol] and !visited[currRow-1][currCol])
            {
                myStack.push({currRow - 1 , currCol}) ;
                visited[currRow - 1][currCol] = true ;
            }
        }
        
        if(validMove(currRow + 1 , currCol , grid.size() , grid[0].size()))
        {
            if(grid[currRow + 1][currCol] and !visited[currRow+1][currCol])
            {
                myStack.push({currRow + 1 , currCol}) ;
                visited[currRow + 1][currCol] = true ;
            }
        }
        
        if(validMove(currRow , currCol - 1 , grid.size() , grid[0].size()))
        {
            if(grid[currRow][currCol - 1] and !visited[currRow][currCol-1])
            {
                myStack.push({currRow , currCol - 1}) ;
                visited[currRow][currCol - 1] = true ;
            }
        }
        
        if(validMove(currRow , currCol + 1 , grid.size() , grid[0].size()))
        {
            if(grid[currRow][currCol+1] and !visited[currRow][currCol+1])
            {
                myStack.push({currRow , currCol + 1}) ;
                visited[currRow][currCol + 1] = true ;   
            }
        }  
    }
    return area ;
}
int Solution :: maxAreaOfIsland(vector<vector<int>>& grid)
{
    maxArea = INT_MIN ;
    int row = grid.size() , col = grid[0].size() ;
    vector<vector<bool>>visited(row , vector<bool>(col , false)) ;
    for(int i = 0 ; i < row ; i++)
    {
        for(int j = 0 ; j < col ; j++)
        {
            if(grid[i][j] and !visited[i][j])
            {
                startRow = i ;
                startCol = j ;
                maxArea = max(maxArea , dfs(grid , visited)) ;
            }
        }
    }
    return maxArea == INT_MIN ? 0 : maxArea ;
}

File: leetcode/practice/695-max-area-of-island/dfs.java
class Solution
{
    boolean[][] visited ;
    int[][] grid ;

    public boolean isValidDir(int i , int j , int r , int c)
    {
        if(i >= 0 && i < r && j >= 0 && j < c)
            return true ;
        return false ;
    }

    public int dfs(int startRow , int startCol , int row , int col)
    {
        Stack<Pair<Integer , Integer>> stk = new Stack<>() ;
        Pair<Integer , Integer> pr = new Pair(startRow , startCol) ;
        stk.push(pr) ;
        int area = 0 ;
        while(!stk.empty())
        {
            pr = stk.peek() ;
            stk.pop() ;
            int r = pr.getKey() ;
            int c = pr.getValue() ;
            // To skip those cells which were already visited but
            // are still present in stack .
            // Eg -- [[1 , 1] , [1 , 1]] -- If we dont add this if
            // statement , then max area will be 5 instead of 4 .
            if(visited[r][c])
                continue ;
            visited[r][c] = true ;
            area++ ;
            if(isValidDir(r , c + 1 , row , col))
                if(grid[r][c + 1] == 1 && !visited[r][c + 1])
                    stk.push(new Pair(r , c + 1)) ;

            if(isValidDir(r , c - 1 , row , col))
                if(grid[r][c - 1] == 1 && !visited[r][c - 1])
                    stk.push(new Pair(r , c - 1)) ;

            if(isValidDir(r + 1 , c , row , col))
                if(grid[r + 1][c] == 1 && !visited[r + 1][c])
                    stk.push(new Pair(r + 1 , c)) ;

            if(isValidDir(r - 1 , c , row , col))
                if(grid[r - 1][c] == 1 && !visited[r - 1][c])
                    stk.push(new Pair(r - 1 , c)) ;
        }
        return area ;
    }
    public int maxAreaOfIsland(int[][] grid)
    {
        visited = new boolean[grid.length][grid[0].length] ;
        this.grid = grid ;
        int area = Integer.MIN_VALUE ;
        int row = grid.length , col = grid[0].length ;
        for(int i = 0 ; i < row ; i++)
            for(int j = 0 ; j < col ; j++)
                if(grid[i][j] == 1 && !visited[i][j])
                    area = Math.max(area , dfs(i , j , row , col)) ;
        return area == Integer.MIN_VALUE ? 0 : area ;
    }
}
